File: backend/app/models/study_session.py
# Geplantes Lerntreffen zwischen zwei gematchten Nutzern.
from uuid import uuid4
from datetime import datetime, timezone
from sqlalchemy import Column, Date, Time, Enum, DateTime, ForeignKey
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship

from app.models.base import Base
from app.models.enums import SessionStatus


class StudySession(Base):
    __tablename__ = "study_sessions"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    match_id = Column(UUID(as_uuid=True), ForeignKey("matches.id"), nullable=False)
    raum_id = Column(UUID(as_uuid=True), ForeignKey("rooms.id"), nullable=True)
    datum = Column(Date, nullable=False)
    uhrzeit = Column(Time, nullable=False)
    status = Column(Enum(SessionStatus, name="session_status_enum", create_type=False), default=SessionStatus.angefragt)
    created_by_id = Column(UUID(as_uuid=True), nullable=True)
    proposed_datum = Column(Date, nullable=True)
    proposed_uhrzeit = Column(Time, nullable=True)
    proposed_raum_id = Column(UUID(as_uuid=True), nullable=True)
    edit_proposed_by_id = Column(UUID(as_uuid=True), nullable=True)
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))

    match = relationship("Match", back_populates="study_sessions")
    room = relationship("Room", back_populates="study_sessions")

File: frontend/lib/shared/models/study_session.dart
class StudySession {
  final String id;
  final String matchId;
  final String datum;
  final String uhrzeit;
  final String? uhrzeitEnde;
  final String status;
  final String? raumId;
  final String partnerAlias;
  final String? createdById;
  final String? proposedDatum;
  final String? proposedUhrzeit;
  final String? proposedUhrzeitEnde;
  final String? proposedRaumId;
  final String? editProposedById;
  final bool iProposedEdit;

  const StudySession({
    required this.id,
    required this.matchId,
    required this.datum,
    required this.uhrzeit,
    this.uhrzeitEnde,
    required this.status,
    required this.partnerAlias,
    this.raumId,
    this.createdById,
    this.proposedDatum,
    this.proposedUhrzeit,
    this.proposedUhrzeitEnde,
    this.proposedRaumId,
    this.editProposedById,
    this.iProposedEdit = false,
  });

  factory StudySession.fromJson(Map<String, dynamic> json) => StudySession(
        id: json['id'] as String,
        matchId: json['match_id'] as String,
        datum: json['datum'] as String,
        uhrzeit: json['uhrzeit'] as String,
        uhrzeitEnde: json['uhrzeit_ende'] as String?,
        status: json['status'] as String,
        raumId: json['raum_id'] as String?,
        partnerAlias: json['partner_alias'] as String? ?? 'Unbekannt',
        createdById: json['created_by_id'] as String?,
        proposedDatum: json['proposed_datum'] as String?,
        proposedUhrzeit: json['proposed_uhrzeit'] as String?,
        proposedUhrzeitEnde: json['proposed_uhrzeit_ende'] as String?,
        proposedRaumId: json['proposed_raum_id'] as String?,
        editProposedById: json['edit_proposed_by_id'] as String?,
        iProposedEdit: json['i_proposed_edit'] as bool? ?? false,
      );

  bool get hasPendingEdit => editProposedById != null;
  bool get isRequested => status == 'angefragt';

  DateTime get dateTime {
    final time = uhrzeit.length > 5 ? uhrzeit.substring(0, 5) : uhrzeit;
    return DateTime.parse('${datum}T$time:00');
  }

  bool get isUpcoming => dateTime.isAfter(DateTime.now());
}
